nomad/fsm.go

```diff
@@ -204,6 +207,49 @@ func (n *nomadFSM) applyUpsertJob(buf []byte, index uint64) interface{} {
 		n.logger.Printf("[ERR] nomad.fsm: UpsertJob failed: %v", err)
 		return err
 	}
+
+	// If it is periodic, insert it into the periodic runner and record the
+	// time it was inserted.
+	if req.Job.IsPeriodic() {
+		if err := n.periodicRunner.Add(req.Job); err != nil {
+			n.logger.Printf("[ERR] nomad.fsm: PeriodicRunner.Add failed: %v", err)
+			return err
+		}
+
+		// Record the insertion time as a launch.
+		launch := &structs.PeriodicLaunch{req.Job.ID, time.Now()}
+		if err := n.state.UpsertPeriodicLaunch(index, launch); err != nil {
+			n.logger.Printf("[ERR] nomad.fsm: UpsertPeriodicLaunch failed: %v", err)
+			return err
+		}
+	}
+
+	// Check if the parent job is periodic and mark the launch time.
+	parentID := req.Job.ParentID
+	if parentID != "" {
+		parent, err := n.state.JobByID(parentID)
+		if err != nil {
+			n.logger.Printf("[ERR] nomad.fsm: JobByID(%v) lookup for parent failed: %v", parentID, err)
+			return err
+		} else if parent == nil {
+			// The parent has been deregistered.
+			return nil
+		}
+
+		if parent.IsPeriodic() {
+			t, err := n.periodicRunner.LaunchTime(req.Job.ID)
```

Comment:
It is less precise but given that the difference in time is negligible and the added complexity I agree. Will be going through all the comments soon